### PR TITLE
Added support for all gels cases

### PR DESCRIPTION
**Pull request by tfalders**
Addresses SWDEV-268195. gels will now supported underdetermined and transposed systems.

### CODE REVIEW

**Comment by cgmb** [comment]
Well, I couldn't do it in a day, but you could! Nice work.

File: library/src/lapack/roclapack_gels.hpp [hunk 1]
```diff
@@ -198,44 +238,159 @@ rocblas_status rocsolver_gels_template(rocblas_handle handle,
...
+            // do the equivalent of trtrs
+            hipLaunchKernelGGL(check_singularity<T>, dim3(batch_count, 1, 1),
+                               dim3(1, check_threads, 1), 0, stream, n, A, shiftA, lda, strideA,
+                               info);
+
+            // save elements of B that will be overwritten by TRSM for cases where info is nonzero
+            hipLaunchKernelGGL((masked_copymat<T, U>), dim3(copyblocksx, copyblocksy, batch_count),
+                               dim3(32, 32), 0, stream, copymat_to_buffer, n, nrhs, B, shiftB, ldb,
+                               strideB, ipiv_savedB, info);
+
```
Comment (jzuniga-amd) [inline library/src/lapack/roclapack_gels.hpp:298]: Just to be consistent with the comments in the code, please add:
"Solve R'Y = B overwriting B with Y (here Y = Q'X)"

File: library/src/lapack/roclapack_gels.hpp [hunk 2]
```diff
@@ -198,44 +238,159 @@ rocblas_status rocsolver_gels_template(rocblas_handle handle,
...
+            // do the equivalent of trtrs
+            hipLaunchKernelGGL(check_singularity<T>, dim3(batch_count, 1, 1),
+                               dim3(1, check_threads, 1), 0, stream, m, A, shiftA, lda, strideA,
+                               info);
+
+            // save elements of B that will be overwritten by TRSM for cases where info is nonzero
+            hipLaunchKernelGGL((masked_copymat<T, U>), dim3(copyblocksx, copyblocksy, batch_count),
+                               dim3(32, 32), 0, stream, copymat_to_buffer, m, nrhs, B, shiftB, ldb,
+                               strideB, ipiv_savedB, info);
+
```
Comment (jzuniga-amd) [inline library/src/lapack/roclapack_gels.hpp:340]: Here also, please add
"Solve LY = B overwriting B with Y (here Y = QX)"

File: library/src/lapack/roclapack_gels.hpp [hunk 3]
```diff
@@ -198,44 +238,159 @@ rocblas_status rocsolver_gels_template(rocblas_handle handle,
...
+            // solve RX = Q'B, overwriting B with X
```
Comment (jzuniga-amd) [inline library/src/lapack/roclapack_gels.hpp:381]: And here, of course, the comment is not accurate; it should read:
"Solve L'X = QB, overwriting B with X"

**Review by jzuniga-amd — APPROVED** [review]
LGTM, Good job!!
Some minor comments below...

**Review by cgmb — APPROVED** [review]
LGTM

File: library/src/lapack/roclapack_gels.hpp [hunk 4]
```diff
@@ -50,6 +52,28 @@ __global__ void masked_copymat(copymat_direction direction,
...
+                              const rocblas_int nrhs,
+                              U B,
+                              const rocblas_int shiftB,
+                              const rocblas_int ldb,
+                              const rocblas_stride strideB,
+                              const rocblas_int* info)
+{
+    const auto b = hipBlockIdx_z;
+    const auto i = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    const auto j = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
```
[the remark this answers is not in the document]
Comment (tfalders) [reply library/src/lapack/roclapack_gels.hpp:67]: That should be easily doable with a search-and-replace. Might be worth doing after Juan's upcoming PR, as he's likely used this style and I have no forthcoming PRs for rocSOLVER.

**Pull request merged**